internal/storage/storage.go

```diff
@@ -0,0 +1,220 @@
+	//
+	// See Backend docs for more details.
+	WriteCAS(ctx context.Context, res *pbresource.Resource, version string) (*pbresource.Resource, error)
+
+	// DeleteCAS performs an atomic CAS (Check-And-Set) deletion of a resource
+	// based on its version. The given version will be compared to what is stored,
+	// and if it does not match, ErrConflict will be returned.
+	//
+	// If the resource does not exist (i.e. has already been deleted) no error will
+	// be returned.
+	//
+	// If a deletion cannot be performed because of a consistency or availability
+	// issue (e.g. when interacting with a Raft follower, or when quorum is lost)
+	// ErrInconsistent will be returned.
+	//
+	// # UIDs
+	//
+	// If the given id's Uid does not match what is stored, the deletion will be a
+	// no-op (i.e. it is considered to be a different resource).
+	//
+	// See Backend docs for more details.
+	//
+	// # GroupVersion
+	//
+	// Delete does not check or refer to the GroupVersion. Resources of the same
+	// Group and Kind are considered equivalent, so requests to delete a resource
+	// using a new GroupVersion will delete a resource even if it's stored with an
+	// old GroupVersion.
+	//
+	// See Backend docs for more details.
+	DeleteCAS(ctx context.Context, id *pbresource.ID, version string) error
+
+	// List resources of the given type, tenancy, and optionally matching the given
+	// name prefix.
+	//
+	// # Tenancy Wildcard
+	//
+	// In order to list resources across multiple tenancy units (e.g. partitions)
+	// pass the Wildcard sentinel value in tenancy fields.
+	//
+	// # GroupVersion
+	//
+	// The resType argument contains only the Group and Kind, to reflect the fact
+	// that resources may be stored in a mix of old and new forms. As such, it's
+	// the caller's responsibility to check the resource's GroupVersion and
+	// translate or filter accordingly.
+	//
+	// # Consistency
+	//
+	// List makes no guarantees about consistency, and may return stale results.
+	List(ctx context.Context, resType UnversionedType, tenancy *pbresource.Tenancy, namePrefix string) ([]*pbresource.Resource, error)
+
+	// WatchList watches resources of the given type, tenancy, and optionally
+	// matching the given name prefix. Upsert events for the current state of the
+	// world (i.e. existing resources that match the given filters) will be emitted
+	// immediately, and will be followed by delta events whenever resources are
+	// written or deleted.
+	//
+	// See List docs for details about Tenancy Wildcard, GroupVersion, and
+	// Consistency.
+	WatchList(ctx context.Context, resType UnversionedType, tenancy *pbresource.Tenancy, namePrefix string) (Watch, error)
+
+	// OwnerReferences returns the IDs of resources owned by the resource with the
+	// given ID. It is typically used to implement cascading deletion.
+	//
+	// # Consistency
+	//
+	// OwnerReferences makes no guarantees about consistency, and may return stale
```

Comment:
Thanks for raising this! I'd originally convinced myself that stale results were fine because by the point the reaper controller would call this method, the owner had been deleted, so we wouldn't allow the creation of any more owned resources - but that's still racy.

I think this might need to be strongly consistent after all.

Reply:
I _think_ the assumptions will still hold.

If you pass in a resource id that contains a uid then only resources owned by that incarnation of the parent resource should be returned. As the original parent has been deleted, no more resources owned by that incarnation will be created. Others owned by a new incarnation may but we shouldn't care.